src/infiniteLoops/detect.ts

```diff
@@ -0,0 +1,362 @@
+
+const runAltErgo: any = require('alt-ergo-modified')
+
+const options = JSON.stringify({
+  answers_with_loc: false,
+  input_format: 'Native',
+  interpretation: 1,
+  unsat_core: true,
+  verbose: false,
+  sat_solver: 'Tableaux',
+  file: 'smt-file'
+})
+
+/**
+ * Checks if the program is stuck in an infinite loop.
+ * @throws InfiniteLoopError if so.
+ * @returns void otherwise.
+ */
+export function checkForInfiniteLoop(
+  stackPositions: number[],
+  state: st.State,
+  functionName: string | undefined
+) {
+  const report = (message: string, type: InfiniteLoopErrorType) => {
+    throw new InfiniteLoopError(functionName, state.streamMode, message, type)
+  }
+  if (hasNoBaseCase(stackPositions, state)) {
+    report('It has no base case.', InfiniteLoopErrorType.NoBaseCase)
+  }
+  // arbitrarily using same threshold
+
+  let circular
+  try {
+    circular = checkForCycle(stackPositions.slice(stackPositions.length - state.threshold), state)
+  } catch (e) {
+    circular = undefined
+  }
+  if (circular) {
+    let message
+    if (circular[0] === circular[1] && circular[0] === '') {
+      message = 'None of the variables are being updated.'
+    } else {
+      message = 'It has the infinite cycle: ' + circular.join(' -> ') + '.'
+    }
+    report(message, InfiniteLoopErrorType.Cycle)
+  } else {
+    const code = codeToDispatch(stackPositions, state)
+    const pass = runUntilValid(code)
+    if (pass) {
+      const message = 'In particular, ' + pass
+      report(message, InfiniteLoopErrorType.FromSmt)
+    }
+  }
+}
+
+/**
+ * If no if statement/conditional was encountered between iterations, there is no base case.
+ */
+function hasNoBaseCase(stackPositions: number[], state: st.State) {
+  const thePaths = state.mixedStack.slice(stackPositions[0], stackPositions[1]).map(x => x.paths)
+  return flatten(thePaths).length === 0
+}
+
+/**
+ * @returns if a cycle was detected, string array describing the cycle. Otherwise returns undefined.
+ */
+function checkForCycle(stackPositions: number[], state: st.State): string[] | undefined {
+  const maybeConc = stackPositions.map(i => state.getMaybeConc(i))
+  const concStr = []
+  for (const item of maybeConc) {
+    const innerStr = []
+    for (const [name, val] of item) {
+      if (typeof val === 'function') {
+        return
+      }
+      innerStr.push(`(${getOriginalName(name)}: ${stringifyCircular(val)})`)
+    }
+    concStr.push(innerStr.join(', '))
+  }
+  return getCycle(concStr)
+}
+
+function getCycle(temp: any[]) {
+  const last = temp[temp.length - 1]
+  const ix1 = temp.lastIndexOf(last, -2)
+  if (ix1 === -1) return undefined
+  const period = temp.length - ix1 - 1
+  const s1 = temp.slice(ix1 - period, ix1)
+  const s2 = temp.slice(ix1, -1)
+  if (s1.length != period) return undefined
+  for (let i = 0; i < period; i++) {
+    if (s1[i] != s2[i]) return undefined
+  }
+  return s1.concat(s1[0])
+}
+
+function stringifyCircular(x: any) {
```

Comment:
I think we discussed not using any kind of stringify at all, but if we're using one, what's wrong with using the standard library's stringify? see `utils/stringify.ts`

Reply:
Didn't know there was already a stringify implementation... In any case I'll just leave this bit in for now until https://github.com/source-academy/js-slang/issues/1074 is resolved.